gitlab/resource_gitlab_project_approval_rule_test.go

```diff
@@ -0,0 +1,416 @@
+					return c.EqualLists(expectedData.Get(attribute), receivedData.Get(attribute))
+				default:
+					return nil // Skipping
+				}
+			}
+			if attrValue.Computed {
+				attrDefault, err := attrValue.DefaultValue()
+				if err != nil {
+					return err
+				}
+				if attrDefault == nil {
+					return nil
+				}
+			}
+			expectedValue := getValueFromGitLabApprovalRule(&expected, attribute)
+
+			return resource.TestCheckResourceAttr(name, attribute, expectedValue)(state)
+		})
+	}
+
+	return resource.ComposeAggregateTestCheckFunc(checks...)
+}
+
+func (c *checkGitlabProjectApprovalRule) EqualLists(expected, received interface{}) error {
+	if expected.(*schema.Set).Len() != received.(*schema.Set).Len() {
+		return fmt.Errorf("Expected set count (%d) but received (%d)", expected.(*schema.Set).Len(), received.(*schema.Set).Len())
+	}
+
+	return nil
+}
+
+func (c *checkGitlabProjectApprovalRule) Exists(name string, rule *gitlab.ProjectApprovalRule) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		var err error
+
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", name)
+		}
+
+		var pid, rid string
+		if pid, rid, err = parseTwoPartID(rs.Primary.ID); err != nil {
+			return err
+		}
+
+		if pid == "" || rid == "" {
+			return fmt.Errorf(`project ID "%s" or rule ID "%s" isn't set`, pid, rid)
+		}
+
+		client := testAccProvider.Meta().(*gitlab.Client)
+		if rules, _, err := client.Projects.GetProjectApprovalRules(pid); err == nil {
+			var ruleID int
+			if ruleID, err = strconv.Atoi(rid); err != nil {
+				return err
+			}
+
+			if len(rules) == 0 {
+				return nil
+			}
+
+			for _, r := range rules {
+				if r.ID == ruleID {
+					*rule = *r
+
+					return nil
+				}
+			}
+		}
+
+		return err
+	}
+}
+
+func (c *checkGitlabProjectApprovalRule) Destroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*gitlab.Client)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != resourceGitLabProjectApprovalRuleName {
+			continue
+		}
+
+		pid, rid, err := parseTwoPartID(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		rules, _, err := conn.Projects.GetProjectApprovalRules(pid)
+		if err == nil && len(rules) == 0 {
+			return nil
+		}
+
+		var ruleID int
+
+		if ruleID, err = strconv.Atoi(rid); err != nil {
+			return err
+		}
+
+		if len(rules) != 0 {
+			for _, rule := range rules {
+				if rule.ID == ruleID {
+					return fmt.Errorf("rule still exists")
+				}
+			}
+		}
+		return nil
+	}
+
+	return nil
+}
+
+// getValueFromGitLabApprovalRule is a mapping to help do a lookup from a rule
+// object via string. Used for testing.
+func getValueFromGitLabApprovalRule(rule *gitlab.ProjectApprovalRule, key string) string {
+	var ret interface{}
+
+	switch key {
+	case "name":
+		ret = rule.Name
+	case "approvals_required":
+		ret = rule.ApprovalsRequired
+	case "group_ids":
+		ret = flattenApprovalRuleGroupIDs(rule.Groups)
```

Comment:
In general can you avoid calling the same helper functions in your test as you do in your main code? You run the risk of testing for the same buggy behavior that you may have introduced. If there are bugs in these functions you will not catch them.